Wireless tags platform

sergeymaysak opened the pull request:
## Description:
This PR introduces support of [Wireless Sensor Tag](http://wirelesstag.net) sensors in HASS.
Each tag depending on its type provides a set of metrics including `temperature`, `humidity`, `light`, `moisture` and `motion`.

Each metric type supplements with switch which allows to `arm/disarm` monitoring of specific sensor type changes. In general, all metrics are updated automatically, but `motion` needs explicit arming procedure using switch.

Each metric type provides extensive set of events represented by this implementation as binary sensor states. It includes `heat` or `cold` temperature conditions, `too wet` and `too dry` humidity conditions, `motion detection` and `door open/close` states, `light` presence as well as water `leak` presence. Supplementary each tag provides `low battery` and `in range/out of range` states.

All binary sensor states are received by defining `local push notification` from tag manager controlling all tags in question. Additionally, there is installed local push notification for each tag for update each metric value event - so temperature or light or humidity values are updated immediately using local traffic. In addition to this, as supplementary measure, `cloud polling` is used to receive initial states of each tag as well as to have ability to get data on demand any time needed. Regardless of frequency of polling attempts, the actual requests to cloud server are performed not earlier as update interval set by end user at tag manager settings in `wirelesstags` account.

The implementation is tested on 3 different types of actual wirelesstag hardware including:

- 13-bit tag - (temperature/humidity/motion) sensor tag type
- water/moisture tag - leak, water/soil moisture level sensor type (includes temperature as well)
- ALS Pro 8-bit memory tag - ambient light sensor type with internal memory for storing data while out of range - includes motion, light, temperature and humidity metrics sensors.
Current implementation is based on [wirelesstagpy](https://github.com/sergeymaysak/wirelesstagpy) (written by me) which uses sync/blocking http networking (based on requests library).

So far it looks like current implementation depends on presence of binary sensors config, otherwise local push is not installed. I'm not sure how to resolve it as I don't know (at the moment) how to determine if specific part of platform (sensors or binary_sensors) were initialized being at main entry to platform setup.

Thank you )

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** https://github.com/home-assistant/home-assistant.github.io/pull/5035

## Example entry for `configuration.yaml` (if applicable):
```yaml
wirelesstag:
  username: you@example.com
  password: secret
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

sergeymaysak committed:
Initial version of wirelesstags platform support.

sergeymaysak committed:
Pinned wirelesstagpy, generated requirements_all.

homeassistant commented:
Hi @sergeymaysak, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#13495).

Once you do that we will be able to review and accept this pull request.

Thanks!

tube0013 commented:
I'd like to test this out.  How do you define local push notifications from the tag manager?

sergeymaysak commented:
@tube0013 - I've use tag's manager API which allows to install custom URL call for specific events in tag. So I define local push notifications as POST request with defined by me payload as json with parameters filled by tag manager. I specify 'nat' parameter as true for tag manager that make them local (so I can use local ip address of hass). To receive these notifications, I define two custom events (and corresponding event handlers) in hass (see homeassistant/components/wirelesstag.py lines 197, 198). 
example of URLs I use look as following: 
 - http://192.168.1.50:8123/api/events/update_tags for temp/humidity/light actual value update
 - http://192.168.1.50:8123/api/events/binary_event for binary sensors events (like door open/closed, water detected)

Please note - to make it working you need to whitelist IP of your tag manager in HAS http configuration. Actually, I've forgot to mention it in documentation, so question is how can I do it now (as my documentation has been already reviewed here [docs pull request](https://github.com/home-assistant/home-assistant.github.io/pull/5035))?

Does it help? ;)

tube0013 commented:
@sergeymaysak thank you, this makes perfect sense. I was not able to get the component to load in my hassio install. I'll keep trying as time allows, I don't think this PR is the place for this continues discussion though. - if you have another place that works better let me know-

I have all sensor/binary_sensor/switch wirelesstag.py files in the corresponding directories under custom_components, and I have the componet wirelesstag.py in the root of custom_components.

`2018-04-03 15:55:37 ERROR (MainThread) [homeassistant.loader] Error loading custom_components.binary_sensor.wirelesstag. Make sure all dependencies are installed
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/homeassistant/loader.py", line 142, in get_component
    module = importlib.import_module(path)
  File "/usr/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/config/custom_components/binary_sensor/wirelesstag.py", line 13, in <module>
    from homeassistant.components.wirelesstag import (
ModuleNotFoundError: No module named 'homeassistant.components.wirelesstag'
2018-04-03 15:55:37 ERROR (MainThread) [homeassistant.loader] Unable to find component binary_sensor.wirelesstag
2018-04-03 15:55:38 ERROR (MainThread) [homeassistant.loader] Error loading custom_components.switch.wirelesstag. Make sure all dependencies are installed
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/homeassistant/loader.py", line 142, in get_component
    module = importlib.import_module(path)
  File "/usr/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/config/custom_components/switch/wirelesstag.py", line 12, in <module>
    from homeassistant.components.wirelesstag import (
ModuleNotFoundError: No module named 'homeassistant.components.wirelesstag'
2018-04-03 15:55:38 ERROR (MainThread) [homeassistant.loader] Unable to find component switch.wirelesstag
2018-04-03 15:55:39 ERROR (MainThread) [homeassistant.loader] Error loading custom_components.sensor.wirelesstag. Make sure all dependencies are installed
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/homeassistant/loader.py", line 142, in get_component
    module = importlib.import_module(path)
  File "/usr/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/config/custom_components/sensor/wirelesstag.py", line 14, in <module>
    from homeassistant.components.wirelesstag import (
ModuleNotFoundError: No module named 'homeassistant.components.wirelesstag'
2018-04-03 15:55:39 ERROR (MainThread) [homeassistant.loader] Unable to find component sensor.wirelesstag`

sergeymaysak commented:
@tube0013 actually, the intent of this PR is to avoid need of custom_component setup :) (from hass 0.67 i hope), but meanwhile to allow your custom setup you need to adjust  `import` directives in  all sub-components (binary_sensor, sensor and switch). I've tried it myself and can confirm it works:
- you need to change in custom_components/binary_sensor, custom_components/sensor, custom_components/switch the line `from homeassistant.components.wirelesstag import` to
`from custom_components.wirelesstag import` to allow custom component t be loaded.
enjoy and any comments/concerns are welcome ;)

tube0013 commented:
@sergeymaysak okay thank you, that got it running.  I do get this error on my frontend: 
`Error: failed to install local push notifications`

the tag manager didn't let me use https for my call back url.  so I'm am I right to assume this won't work locally if my HA instance is using https?

sergeymaysak commented:
@tube0013 superb observation! need to check it. My hass is in http (I use `homebridge` to access remotely to it's parts ;) ). What would make it 's'? presence of tls certificate in http component? any info on your config could help )

sergeymaysak commented:
@tube0013 my tests with my Postmap setup shows that Wireless tag returns `"Tag Manager only supports http://... URL"` for attempt to install https:// based scheme url. So it sounds as, yes http only is allowed.

tube0013 commented:
@sergeymaysak yes once you add the certificates to the http component HASS is now over https.  I reconfigured my set up to put caddy in front of hass and handling the ssl, so hass can be accessed over http from inside my lan now.  - how can I be sure local updates from the tag manager are working?

Also wirelesstag is reporting temperature values in Celsius only, HA can do the conversion to F if set to imperial units.  The rest api sends the temps in C as well, and HA displays them in F for me.  Your code looks to be assuming they are coming in F.

<img width="882" alt="screen shot 2018-04-04 at 11 40 55 pm" src="https://user-images.githubusercontent.com/11084412/38346230-b873ddcc-3861-11e8-9060-986f321f8b3d.png">

<img width="383" alt="screen shot 2018-04-04 at 11 42 12 pm" src="https://user-images.githubusercontent.com/11084412/38346254-da553a30-3861-11e8-8c85-43d8ec794ce9.png">

so the temp is showing as 20.1 F but is really 20.1 C, and should read ~68.05 F

tube0013 commented:
@sergeymaysak one other observation now that I mostly got this running, HA doesn't seem to like it if there is white space in the Tag names it gets from wirelesstag.  I was getting errors about invalid entity names.  I replaced all the space with underscores in my tag names, and now I'm not seeing the errors.  Here is an example of the error I was getting:

`Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/homeassistant/helpers/entity_platform.py", line 245, in _async_add_entity
    'Invalid entity id: {}'.format(entity.entity_id))
homeassistant.exceptions.HomeAssistantError: Invalid entity id: sensor.wirelesstag_Laundry Room_light`

sergeymaysak committed:
Fixed temperature units in imperial units systems, make binary events more tags specific.

sergeymaysak committed:
Lowercased tag name during entity_id generation.

sergeymaysak commented:
@tube0013 Thanks for observations! I've updated my code to proper handling of unit_system: imperial. Additionally I've updated name of `binary_event` to be `tag_binary_event` to make sure it's name does not clash with any other platforms. So please take all 4 files again from this PR and re-iterate changes for `import` for your custom_components. 
`how can I be sure local updates from the tag manager are working` - you can observe in console the message `Push notification for update arrived:`...
As for tags names - very strange as most of my tags are containing whitespaces  - `Kids room Water` for example and I have dedicated code which replaces all whitespaces with underscore for sensors. So not sure how you get `sensor.wirelesstag_Laundry Room_light` as entity id (.

tube0013 commented:
@sergeymaysak thank you, I finally had some time to update the component files and re-test, and everything seems to be working well!

sergeymaysak committed:
Fixed: tag_id template for tag_binary_events, support of light sensor for homebridge.

sergeymaysak commented:
@tube0013 I've identified critical defects in handing of binary events and fixed it. Please get update if you are interested in ). Additionally I've found and fixed defect which prevented integration of light sensor with Homebridge.

sergeymaysak committed:
Minor style cleanup.

sergeymaysak commented:
@andrey-git @balloob @frenck Guys, is there any chance to be reviewed and merged? ;)

balloob commented:
Please do not tag people asking for reviews. There are a lot of open reviews. We'll get to it when we get to it.

sergeymaysak commented:
@balloob got you. sorry for being inpatient and hope if it does not look rude from my side. i really appreciate your amazing work here.

balloob commented on homeassistant/components/switch/wirelesstag.py:
```diff
@@ -0,0 +1,129 @@
+}
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_ENTITY_NAMESPACE, default=DEFAULT_ENTITY_NAMESPACE):
```
This is already part of PLATFORM_SCHEMA.

sergeymaysak replied:
@balloob got it. removing...

balloob commented on homeassistant/components/switch/wirelesstag.py:
```diff
@@ -0,0 +1,129 @@
+        return False
+
+    @property
+    def state(self):
```
Do not override `state`, just implement `is_on` property

sergeymaysak replied:
@baloob I have state defined in my base class, so if I do not override state here, switch does not work as expected.

sergeymaysak replied:
got it. reworking...

balloob commented on homeassistant/components/switch/wirelesstag.py:
```diff
@@ -0,0 +1,129 @@
+    @property
+    def assumed_state(self):
+        """Return true if unable to access real state of entity."""
+        return False
```
this is the default, you can remove this property.

sergeymaysak replied:
@balloob thanks! removing it...

balloob commented on homeassistant/components/binary_sensor/wirelesstag.py:
```diff
@@ -0,0 +1,208 @@
+        self._tag_attr = SENSOR_TYPES[self._sensor_type][2]
+        self.binary_spec = SENSOR_TYPES[self._sensor_type][3]
+        self.tag_id_index_template = SENSOR_TYPES[self._sensor_type][4]
+        self._state = self.updated_state_value()
```
Don't do this. Instead, you're already passing `True` to `add_devices`, so Home Assistant will call the update method.

sergeymaysak replied:
@balloob If concern is to not load any data here, but let hass call update method where some network activity will be done, then self.updated_state_value() do not load actual data. As it name states - it returns updated value - i.e current known value from last update.

sergeymaysak replied:
ok, removing it...

balloob commented on homeassistant/components/binary_sensor/wirelesstag.py:
```diff
@@ -0,0 +1,208 @@
+    add_devices(sensors, True)
+
+
+class WirelessTagBinarySensor(WirelessTagBaseSensor, BinarySensorDevice):
```
You probably want to implement the `device_class` property so that they get the right icons in the frontend.

sergeymaysak replied:
got it. in progress

sergeymaysak replied:
done.

balloob commented on homeassistant/components/binary_sensor/wirelesstag.py:
```diff
@@ -0,0 +1,208 @@
+        """Handle push notification sent by tag manager."""
+        pass
+
+    def on_binary_event(self, event):
```
If we get push notifications, you should have the `should_poll` property return False.

sergeymaysak replied:
@balloob Well, I actually have both - local push notifications and cloud polling in place. As it stated in description of [`Classification of the Internet of Things`](https://www.home-assistant.io/blog/2016/02/12/classifying-the-internet-of-things/#classifiers) Disadvantage of `local push` is `If it does not also support polling, home automation will not be made aware of the state after booting up until it changes.` So I do support both ways supplementing each others. Actual network activity is not happen every time update is called - it is controlled by polling interval which is controlled and defined by end user at her my.wirelesstag.net account. lib respects this and keep cache of last updated values.
Additionally, there are situations present when local push is not installed - in particular if end user have not defined at least one binary sensor to have push notifications registered (I've warned on it in supplementary documentation here https://github.com/home-assistant/home-assistant.github.io/pull/5035)

balloob replied:
The first part could be circumvented by only polling the cloud once at startup, after that just rely on the pushes to come in.

If you control polling yourself (using `hass.helpers.async_track_interval`), you could disable polling whenever push is active and disable it otherwise. We wouldn't want to add unnecessary usage on the wireless tag API.

sergeymaysak replied:
I do not control polling myself relying on hass to do it for me. There are no unnecessary network traffic happens as library effectively manages time to update using polling interval defined by end user inside wireless tags hw - any calls of library in between polling interval returns cached data. Also I think keeping polling in place make solution more robust - I could imagine scenario when tags reconfigured outside of hass on other url calls which would just stop any push notifications to hass. So polling would continue to provide data... Does it make sense from your prospective?

balloob commented on homeassistant/components/sensor/wirelesstag.py:
```diff
@@ -0,0 +1,168 @@
+SENSOR_MOISTURE = 'moisture'
+SENSOR_LIGHT = 'light'
+
+SENSOR_TEMPERATURE_F_ICON = 'temperature-fahrenheit'
```
Don't define icons. Instead use device classes for sensors.

sergeymaysak replied:
got it. in progress.

sergeymaysak replied:
icons removed.

balloob commented on homeassistant/components/sensor/wirelesstag.py:
```diff
@@ -0,0 +1,168 @@
+    platform = hass.data.get(WIRELESSTAG_DOMAIN)
+    sensors = []
+    tags = platform.tags
+    for _, tag in tags.items():
```
`for tag in tags.values():`

sergeymaysak replied:
thanks. ok )

sergeymaysak replied:
done.

balloob commented on homeassistant/components/sensor/wirelesstag.py:
```diff
@@ -0,0 +1,168 @@
+        self._tag_attr = SENSOR_TYPES[self._sensor_type]['attr']
+        self._unit_of_measurement = SENSOR_TYPES[self._sensor_type]['unit']
+        self._icon = 'mdi:{}'.format(SENSOR_TYPES[self._sensor_type]['icon'])
+        if sensor_type == SENSOR_TEMPERATURE and not config.units.is_metric:
```
Don't do this. Home Assistant will convert all values to the user preferred value. We should not set icons generally, instead rely on device_class to do this.

sergeymaysak replied:
done.

balloob commented on homeassistant/components/sensor/wirelesstag.py:
```diff
@@ -0,0 +1,168 @@
+        self.define_name(self._tag.name)
+
+        # sensor.wirelesstag_bedroom_temperature
+        self._entity_id = '{}.{}_{}_{}'.format('sensor', WIRELESSTAG_DOMAIN,
```
Why would you define an entity id? Way better to define a unique ID if you can and allow the user to manage entity id via the entity registry.

balloob commented on homeassistant/components/sensor/wirelesstag.py:
```diff
@@ -0,0 +1,168 @@
+        self._entity_id = '{}.{}_{}_{}'.format('sensor', WIRELESSTAG_DOMAIN,
+                                               self.underscored_name,
+                                               self._sensor_type)
+        self._state = self.updated_state_value()
```
Don't do this. Passing `True` to `add_devices` will take care of this.

sergeymaysak replied:
ok, removing it. please note however - `updated_state_value` just returns already received value, there is no network loading or something...

sergeymaysak replied:
done.

balloob commented on homeassistant/components/sensor/wirelesstag.py:
```diff
@@ -0,0 +1,168 @@
+        return getattr(self._tag, self._tag_attr, False)
+
+    def update_tag_info(self, event):
+        """Handle push notification sent by tag manager."""
```
Add `should_poll` property that returns False.

balloob commented on homeassistant/components/switch/wirelesstag.py:
```diff
@@ -0,0 +1,129 @@
+        """Initialize a switch for Wireless Sensor Tag."""
+        super().__init__(api, tag)
+        self._switch_type = switch_type
+        self.sensor_type = SWITCH_TYPES[self._switch_type][1]
```
sensor or switch?

sergeymaysak replied:
It is a switch to arm specific sensor type monitoring.  so sensor_type property describe exactly this - what sensor type to monitor as result or arming this switch. - like arm motion or arm temperature or arm humidity monitoring...

sergeymaysak commented on homeassistant/components/sensor/wirelesstag.py:
```diff
@@ -0,0 +1,168 @@
+}
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_ENTITY_NAMESPACE, default=DEFAULT_ENTITY_NAMESPACE):
```
@balloob got it. removing...

sergeymaysak replied:
done.

sergeymaysak commented on homeassistant/components/wirelesstag.py:
```diff
@@ -0,0 +1,299 @@
+                config = self.make_push_notitication(name, binary_url, content)
+                configs.append(config)
+
+        content = ("{\"name\":\"{0}\",\"id\":{1},\"temp\":{2}," +
```
@balloob Well, `content` is very customizable (i.e can be any json), so putting specifically what I've defined as reasonable payload for hass into library, might not be appropriate for any other clients of library. So to the moment, I've not found better balance between hiding complexity and keeping it not dedicated or limited to specific client (i.e keep it general purpose) as what you've seen in my code. What is really protocol specific here is binding values as {0} or {1}, but I'm sure how to solve it - i.e how to hide them inside library but allow to customize content at the same time...

balloob replied:
I don't really follow. Aren't name, id and temp things that the wireless tag API understands?

sergeymaysak replied:
name, id and temp are defined by specifically for hass. tag API does not care how I name keys for values and which keys-value pairs used. it specifies `placeholders` that i could use in my payload template. `{0}` is a placeholder for name of tag, `{2}` for current temperature in C etc. I can define any json payload i want using these placeholders - for example as array {values: [{0}, {2}]} or as I defined for hass as normal JS object.

balloob replied:
Okay, then this is fine.

sergeymaysak commented on homeassistant/components/wirelesstag.py:
```diff
@@ -0,0 +1,299 @@
+        """Factory for notification config."""
+        from wirelesstagpy import NotificationConfig
+        return NotificationConfig(name, {
+            'url': url, 'verb': 'POST',
```
@balloob I'm trying to keep balance between hiding complexity and specifics of protocol used and attempt to make sure library is generic enough to be useful not only in hass, but by other clients as well... Maybe I need more time to come to something better that I've designed now...

balloob replied:
that's what confuses me, wouldn't any other client that needs to send a notification use the same body but just different parameters?

sergeymaysak replied:
If you mean that any clients would need to specify url, verb, content, disabled and nat parameters to properly setup NotificationConfig instance, then yes all these parameters are needed and can be different for different clients. So exactly the need to allow different clients to specify different parameters forces me to keep them all as part of public API. Could you please describe your vision in more details? What exactly do you think could go to library? Some of these parameters? or NotificationConfig class?

sergeymaysak commented on homeassistant/components/wirelesstag.py:
```diff
@@ -0,0 +1,299 @@
+        return self._name
+
+    @property
+    def state(self):
```
got it. reworking...

sergeymaysak replied:
done.

sergeymaysak commented on homeassistant/components/wirelesstag.py:
```diff
@@ -0,0 +1,299 @@
+        """Return the state of the sensor."""
+        return self._state
+
+    def define_name(self, new_name):
```
ok. dropping...

sergeymaysak replied:
done.

sergeymaysak commented on homeassistant/components/wirelesstag.py:
```diff
@@ -0,0 +1,299 @@
+            ATTR_VOLTAGE: '{:.2f}V'.format(self._tag.battery_volts),
+            ATTR_TAG_TYPE: self._tag.tag_type,
+            ATTR_TAG_COMMENT: self._tag.comment,
+            ATTR_TAG_IS_ALIVE: self._tag.is_alive,
```
thanks! defining `available`...

sergeymaysak replied:
done.

sergeymaysak commented on homeassistant/components/wirelesstag.py:
```diff
@@ -0,0 +1,299 @@
+        return False
+
+    # listen to custom event
+    hass.bus.listen('update_tags', hass.data[DOMAIN].handle_update_tags_event)
```
I do not use events to communicate internally between component and platforms. I use them to receive local push rest call.

sergeymaysak replied:
updating to contain component name...

sergeymaysak replied:
done.

sergeymaysak replied:
Signals are used now to communicate from platform to components.

sergeymaysak commented on homeassistant/components/wirelesstag.py:
```diff
@@ -0,0 +1,299 @@
+
+    # listen to custom event
+    hass.bus.listen('update_tags', hass.data[DOMAIN].handle_update_tags_event)
+    hass.bus.listen('tag_binary_event', hass.data[DOMAIN].handle_binary_event)
```
updating to contain component name... (i listen to custom events to define rest call from wirelesstag hw to endpoint of event)

sergeymaysak replied:
done.

sergeymaysak commented on homeassistant/components/wirelesstag.py:
```diff
@@ -0,0 +1,299 @@
+                self._tag.signal_straight),
+            ATTR_TAG_BEEP_DURATION: self._tag.beep_option,
+            ATTR_TAG_OUT_OF_RANGE: not self._tag.is_in_range,
+            ATTR_TAG_HW_REVISION: self._tag.hw_revision,
```
got you. the only static attributes there are ATTR_TAG_HW_REVISION, ATTR_TAG_FW_VERSION and ATTR_TAG_TYPE. Please note - ATTR_TAG_FW_VERSION can be potentially upgraded as well. Any other attributes are very dynamic.

sergeymaysak replied:
removed static attributes.

sergeymaysak commented on homeassistant/components/wirelesstag.py:
```diff
@@ -0,0 +1,299 @@
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return {
```
@balloob Would you agree if I will remove only static attributes? i.e ATTR_TAG_HW_REVISION, ATTR_TAG_FW_VERSION and ATTR_TAG_TYPE are static. 

Others are very dynamic:
- ATTR_VOLTAGE - actual battery voltage - dynamically changing and can be helpful for end user to identify is battery (which is replaceable) needs to be replaced. (automated as well)
- ATTR_TAG_COMMENT - a user defined description of specific tag. can be changed by end user...
- ATTR_TAG_SIGNAL_STRAIGHT - obviously very helpful for end user -allows to diagnose many of issues (no signal etc) - end user is able to change frequency used and other hw settings using this parameter - again it is a subject for automation or say be a trigger for automation
- ATTR_TAG_BEEP_DURATION - user is able to change it - there are about 15 options how tag should beep. I think possibly to add `service` to beep from hass and to allow control beep duration from hass  - in future versions...
- ATTR_TAG_OUT_OF_RANGE - again - by nature - it is very dynamic and very helpful for end user to see if tag is out of range  - tag can be in vehicle or in bag of your kids playing in yard... you can setup notifications on in range, out of range....
- ATTR_TAG_POWER_CONSUMPTION - this attributes changes depending on location of device (it's ATTR_TAG_SIGNAL_STRAIGHT) and allow to determine what would be battery life... very recommended to be shown to end user  - i use it myself to make sure specific tag is located in right way

balloob replied:
Just because it's user configurable doesn't mean that it's not mainly static. So comment should go. Out of range should be used to control the `available` property. Beep duration is a configuration option, we shouldn't expose. 

Home Assistant goal is not to be become a full replacement UI for controlling and configuring wireless tags (or any other IoT device for that matter).

sergeymaysak replied:
ok ) comment will be removed. as `out of range` - i use `is alive` to control `available` as pointed in previous review round. Do you propose to drop `is alive` in favor of `out of range` or combine them? 
IMO `is alive` closer to `available`...

sergeymaysak replied:
comment and beep duration are removed.

balloob replied:
If the tag is out of range, how would you know if it's alive? I think that available should be a combination of both ?

sergeymaysak replied:
Good question ). Info is provided not by tag, but by `tag manager` (a device connected by ethernet to local network) which communicates to actual tags that could be quite distant (up to about 300+ meters). I'm not sure how tag manager provides `alive` different from `out of range`. It is not really difficult technically to combine them. However, I think that guys from `wirelesstags` defined these two properties as different states for a reason - for example - it could be `in range` but so low battery that make it `not alive`... So I think `is alive` is an umbrella state for availability ie - it seems to me reasonable to: 
- provide property `availability` mapped to `is alive`
- keep `out of range` reported as attribute of entity

Would you agree?

sergeymaysak replied:
as side note - tags can be used to record what environmental conditions where during parcels transportation (by putting tag into track). tags are battery operated and some of them have internal memory which is used to record temperature/humidity/light etc so tag can be alive but `out of range` of tag manager. data could be read and analyzed when tag became in range of specific tag manager. Also the memory could save was it `alive` or not during trip.

sergeymaysak commented on homeassistant/components/wirelesstag.py:
```diff
@@ -0,0 +1,299 @@
+        arm_func = getattr(self.api, func_name)
+        if arm_func is not None:
+            arm_func(switch.tag_id)
+            switch.async_schedule_update_ha_state()
```
got it. removing `async_schedule_update_ha_state`

sergeymaysak committed:
Removed state, define_name, icon. Reworked async arm/disarm update. Removed static attrs. Introduced available property. Custom events contains components name now. Cleaned dedundant items from schema definition.

sergeymaysak committed:
Removed comment and beep duration from attributes. Minor cleanup of documentation comment.

sergeymaysak committed:
Ignoring Wemo switches linked in iOS app.

balloob commented on homeassistant/components/wirelesstag.py:
```diff
@@ -0,0 +1,290 @@
+        return False
+
+    # listen to custom event
+    hass.bus.listen('wirelesstag_update_tags',
```
We should use the `helpers.dispatcher` helper to communicate data between components and platforms.

sergeymaysak replied:
I really appreciate your continuous attention to quality of my code. I really tries to conform. To do so I need some `best practices` or example how it should be done. If you could point me to reference implementation or some documentation with example if usage - it would help. Specifically, I dont understand here what exactly should be changed. I use custom event (with changed name to per your request to contain platform name) to receive local push notification from wirelesstag manager (hw bridge). Since platform is my central point hosting all tags and keeping track who needs to get which event I do dispatch myself. Not really clear what should I delegate to `helpers.dispatcher`.

sergeymaysak replied:
Done.

balloob commented on homeassistant/components/binary_sensor/wirelesstag.py:
```diff
@@ -0,0 +1,209 @@
+        if event_type == self.device_class:
+            # state should be 'on' or 'off'
+            self._state = event.data.get('state')
+            self.async_schedule_update_ha_state()
```
This is not called from within an async context. Should be `self.schedule_update_ha_state()`

sergeymaysak replied:
oh! just realized meaning of async_ prefix here. thanks! reworking...

sergeymaysak replied:
I've reworked it along with switch to signals for communication from platform to components.

sergeymaysak committed:
Reworked passing data from platform to components using signals.

sergeymaysak commented on homeassistant/components/wirelesstag.py:
```diff
@@ -0,0 +1,290 @@
+        binary_url = self.binary_event_callback_url
+        for event in self.binary_sensors:
+            for state, name in event.binary_spec.items():
+                content = ('{"type": "' + event.device_class +
```
this passed to api as `template` string, api creates from it json when sends actual event back.

tube0013 commented:
@sergeymaysak question RE local tag URL calling.

I se the url calling on the wireless tag site to use my local ip address. for sensors:

http://192.168.1.xxx:8123/api/events/update_tags

and binary sensors:

http://192.168.1.xxx:8123/api/events/binary_event

this works for a few hours, then I go back and check and the urls have been re-written to my outside domain name which will not work over http...

All sensors get changed to: 

http://DOMAIN:8123/api/events/wirelesstag_update_tags

Some Binary Sensors get changed to:

http://DOMAIN:8123/api/events/wirelesstag_binary_event

I'm not sure why this is happening, and know it makes the local url calling not work.

MartinHjelmare commented:
Please take support questions in our forum or chat. If you suspect a bug, please open an issue. We don't want support or discussion in closed PRs.
